_posts/2015-10-20-jenkins-followup.md

```diff
@@ -0,0 +1,70 @@
+The image already provides out-of-the-box support for installing additional plugins by just providing a list of plugins and versions,
+which we previously did in a much more complicated way using Jenkins CLI.
+
+In addition, the Docker image hinted us to the solution of the configuration problem: Groovy init hooks.
+With these scripts, which utilize the internal API of Jenkins itself and also the plugins, we were able to kick out the precreated `config.xml`.
+As an example, here is a code snippet that creates an account:
+
+```groovy
+    import jenkins.model.*
+    import hudson.security.*
+
+    def instance = Jenkins.getInstance()
+
+    def hudsonRealm = instance.getSecurityRealm()
+
+    def user = hudsonRealm.createAccount("admin", "password")
+    user.setFullName("Admin")
+    user.save()
+```
+
+As you can easily imagine, these scripts add tremendous flexibility to the way of configuring Jenkins, especially when combining the possibilities of Groovy with the dynamic generation of
+these scripts through templates from Ansible.
+
+## Immutability
+
+To get a truly immutable server, there are two things left: Once again, the handling of job metadata, which needs to be persisted, and the Jenkins Web UI, which still allows changing the configuration,
+although these cahnges will be overwritten with the next run of Ansible.
+
+The job metadata problem is easily solved using a standard Docker feature: volumes.
+We are intentionally not mounting the whole `jenkins_home` folder as a persistent volume, but only the reallly important subfolders `jobs` and `nodes`, which contain the actual metadata.
+Like described before, we only backup a small fraction of the files in these folders, as for example the `jobs` folder also contains the workspace,
+which can be created whenever needed by just checking out from version control.
+
+The problem of possible modifications through the web UI can be solved using the [Matrix Authorization Strategy Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Matrix+Authorization+Strategy+Plugin).
+Utilising this plugin, we create only one real admin user, which has a (unknown) random password, and other users, which are only allowed to trigger builds.
+This way, the only way to change the configuration of the server is to create a new version of the Docker container with Ansible.
+Also, jobs can also only be created or modified by the [Job DSL Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Job+DSL+Plugin).
+
+# Outlook
+
+Regarding the general setup of the Jenkins master, we are now more or less where we want to be.
+There are always improvements to make, but the main challanges that lie ahead of us are more in the area of getting our build slaves automated the way we want,
+and running all of our build jobs in an appropriate environment.
+In the end, this is typical everyday CI maintenance, which we wont be able to get rid of completely even with full automation... ;-)
```

Comment:
For internal reasons? The usage of the image is aligned with their legal terms (see https://www.docker.com/brand-guidelines), and it is taken from their own media kit...